opm/parser/eclipse/Parser/tests/MessageContainerTest.cpp

```diff
@@ -0,0 +1,41 @@
+/*
+  Copyright 2016 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#define BOOST_TEST_MODULE MessageContainerTest
+
+#include <boost/test/unit_test.hpp>
+
+#include <opm/parser/eclipse/Parser/MessageContainer.hpp>
+
+
+using namespace Opm;
+
+BOOST_AUTO_TEST_CASE(TestIterator) {
+    
+    MessageContainer msgContainer;
+    msgContainer.error("This is an error.");
+    msgContainer.bug("This is a bug.", "dummy.log", 20);
+    {       
+        BOOST_CHECK_EQUAL("This is an error.", msgContainer.begin()->message);
+        BOOST_CHECK_EQUAL("dummy.log", (msgContainer.end()-1)->location->filename);
+        BOOST_CHECK_EQUAL(20, (msgContainer.end()-1)->location->lineno);
+    }
+    
```

Comment:
> I guess this needs the overload implementation of == and !=, right?

Well; my initial guess was that the default operator== would do; but I guess the `std::unique_ptr<>` creates problems? Come to think of it - that probably creates problems also for the `msgList.push_back( )` - the `std::shared_ptr<T>` does have *some* charm!

Anyway - the important point was that `for( const auto& msg : msgContainer)` should work, the `msgList` was just a (broken ...) way to test it. If you instead test it with:
```C++

std::vector<string> msgList; 
for (const auto& msg : msgContainer) {
    msgList.push_back( msg.message );
}

for (size_t index = 0; index msgList.size(); index++) {
     // requires operator[] - which is a good thing I guess + MessageContainer.size() ??
    const auto& msg = messageContainer[index]; 
    BOOST_CHECK_EQUAL( msgList[index] , msg.message );
}
```

I'm going offline now ;-)